dlib/matrix/matrix_fft.h

```diff
@@ -263,6 +264,291 @@ namespace dlib
+    /*!
+        ensures:
+            - This returns a function object with signature double(size_t i, size_t N) which computes
+              a periodic Kaiser window which can be passed to the STFT family of functions.
+    !*/
+    {
+        return [=](std::size_t i, std::size_t N){return kaiser(i, N, beta, PERIODIC);};
+    }
+
+// ----------------------------------------------------------------------------------------
+
+    template <typename EXP, typename WINDOW>
+    auto stft (
+        const matrix_exp<EXP>& signal,
+        const WINDOW& w,
+        std::size_t fftsize,
+        std::size_t wlen,
+        std::size_t hoplen
+    )
+    {
+        return details::stft_impl(signal, w, fftsize, wlen, hoplen, details::fft_func{});
+    }
+
+// ----------------------------------------------------------------------------------------
+
+    template <typename T, typename Alloc, typename WINDOW>
+    auto stft (
+        const std::vector<T, Alloc>& signal,
+        const WINDOW& w,
+        std::size_t fftsize,
+        std::size_t wlen,
+        std::size_t hoplen
+    )
+    {
+        return stft(dlib::mat(signal), w, fftsize, wlen, hoplen);
+    }
+
+// ----------------------------------------------------------------------------------------
+
+    template <
+        typename EXP,
+        typename WINDOW,
+        typename T = typename EXP::type
+    >
+    auto istft (
+        const matrix_exp<EXP>& stft,
+        const WINDOW& w,
+        std::size_t wlen,
+        std::size_t hoplen
+    )
+    {
+        return details::istft_impl<T>(stft, w, wlen, hoplen, details::ifft_func{});
+    }
+
+// ----------------------------------------------------------------------------------------
+
+    template <typename EXP, typename WINDOW>
+    auto stftr (
+        const matrix_exp<EXP>& signal,
+        const WINDOW& w,
+        std::size_t fftsize,
+        std::size_t wlen,
+        std::size_t hoplen
+    )
+    {
+        return details::stft_impl(signal, w, fftsize, wlen, hoplen, details::fftr_func{});
+    }
+
+// ----------------------------------------------------------------------------------------
+
+    template <typename T, typename Alloc, typename WINDOW>
+    auto stftr (
+        const std::vector<T, Alloc>& signal,
+        const WINDOW& w,
+        std::size_t fftsize,
+        std::size_t wlen,
+        std::size_t hoplen
+    )
+    {
+        return stftr(dlib::mat(signal), w, fftsize, wlen, hoplen);
+    }
+
+// ----------------------------------------------------------------------------------------
+
+    template <
+        typename EXP,
+        typename WINDOW,
+        typename T = typename EXP::type,
+        typename R = remove_complex_t<T>
+    >
+    auto istftr (
+        const matrix_exp<EXP>& stft,
+        const WINDOW& w,
+        std::size_t wlen,
+        std::size_t hoplen
+    )
+    {
+        return details::istft_impl<R>(stft, w, wlen, hoplen, details::ifftr_func{});
+    }
```

Comment:
```suggestion
    template <
        typename EXP,
        typename WINDOW
    >
    auto istftr (
        const matrix_exp<EXP>& stft,
        const WINDOW& w,
        std::size_t wlen,
        std::size_t hoplen
    )
    {
        using R = remove_complex_t<typename EXP::type>;
        return details::istft_impl<R>(stft, w, wlen, hoplen, details::ifftr_func{});
    }
```
These things don't need to be template arguments.  It's simpler this way since these details don't need to be scoped somewhere client code can mess with.

Reply:
Yep